### PR TITLE
feat: MetricsChecker now exposed as a lib

### PR DESCRIPTION
**Pull request by Ray-Eldath**
feat: MetricsChecker now exposed as a lib
test: usage based on test-infra sdk and MetricsChecker lib
bugfix: compensate WaitGroup when errored

### CODE REVIEW

**Comment by Ray-Eldath** [comment]
@cosven ptal :wink:

File: tests/test_with_sdk/.env [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# This file was generated by github.com/pingcap/test-infra/services/rms/cmd/testbed create
+# GENERATED BY THE COMMAND ABOVE; DO NOT EDIT
+TI_PARAM_RES_tc: {"provider":"github.com/pingcap/test-infra/sdk/resource/impl/k8s.TiDBClusterClient","name":"tc","testbed":"testbed-txkds","type":"TIDB_CLUSTER","ready":true}
```
Comment (cosven) [inline tests/test_with_sdk/.env]: `.env` and `output` file should be ignored.

File: pkg/metrics/rules.go [hunk 2]
```diff
@@ -2,23 +2,25 @@ package metrics
...
 
 // Rule represents a bool PromQL expression, returning True means rule satisified.
 //   Send alerts when a rule return true, which is similarly as prometheus
 //   alerting rules.
 type Rule struct {
-	Tag        string `yaml:"tag"`
-	PromQL     string `yaml:"promql"`
-	NotifyFunc NotifyFunc
-	AlertFunc  AlertFunc
+	Tag               string `yaml:"tag"`
```
Comment (cosven) [inline pkg/metrics/rules.go]: Is it proper to rename the `Tag` to `Name`?

File: pkg/metrics/checker.go [hunk 3]
```diff
@@ -5,33 +5,54 @@ import (
...
+		Tag:               tag,
+		PromQL:            promQL,
+		Interval:          interval,
+		EvaluatedCallback: notify,
+		AlertCallback:     alert,
+	})
+	return m
+}
+
+func (m *Checker) AddRule(tag string, interval time.Duration, promQL string) *Checker {
```
Comment (cosven) [inline pkg/metrics/checker.go]: We can add a rule instance factory `CreateRule/NewRule` and pass the `Rule` instance as the parameter.

File: cmd/root.go [hunk 4]
```diff
@@ -36,16 +36,19 @@ func root(cmd *cobra.Command, args []string) {
...
-
-	metricsChecker, err := metrics.NewChecker(Flag.prometheusAPIURL, Conf.Rules, Conf.Interval)
+	metricsChecker, err := metrics.NewChecker(Flag.prometheusAPIURL)
 	if err != nil {
 		log.Fatal(err.Error())
 	}
-	err = metricsChecker.Run()
+	if metricsChecker == nil {
+		return
+	}
```
Comment (cosven) [inline cmd/root.go:45]: It seems we don't need to check if it is nil here.

File: pkg/metrics/checker.go [hunk 5]
```diff
@@ -5,33 +5,54 @@ import (
...
 
+func (m *Checker) AddRuleWithCustomCallback(tag string, interval time.Duration, promQL string, notify EvaluatedCallback, alert AlertCallback) *Checker {
+	m.Rules = append(m.Rules, Rule{
+		Tag:               tag,
+		PromQL:            promQL,
+		Interval:          interval,
+		EvaluatedCallback: notify,
+		AlertCallback:     alert,
+	})
+	return m
```
Comment (cosven) [inline pkg/metrics/checker.go]: Maybe add a comment why return the checker itself.

File: cmd/root.go (hunk 4 above)
Comment (Ray-Eldath) [reply cmd/root.go:45]: this only intend to make GoLand checker happy with `cmd/root.go:48` and `cmd/root.go:51`.

**Comment by Ray-Eldath** [comment]
problems addressed.

btw I think we should have a guideline regarding where to retrieve these metrics. e.g. **always check for [HTTP API](https://github.com/pingcap/tidb/blob/master/docs/tidb_http_api.md) first**, then write your own PromQL.

**Review by cosven — APPROVED** [review]
LGTM

**Pull request merged**